x/ecocredit/server/core/msg_server.go

```diff
@@ -0,0 +1,1200 @@
+		Retired:  toRetiredBalance.String(),
+	}); err != nil {
+		return err
+	}
+
+	// update the "from" balance
+	if err := s.stateStore.BatchBalanceStore().Update(ctx, &ecocreditv1beta1.BatchBalance{
+		Address:  from,
+		BatchId:  batch.Id,
+		Tradable: fromTradableBalance.String(),
+		Retired:  fromRetiredBalance.String(),
+	}); err != nil {
+		return err
+	}
+	// update the "retired" balance only if credits were retired
+	if didRetire {
+		if err := s.stateStore.BatchSupplyStore().Update(ctx, &ecocreditv1beta1.BatchSupply{
+			BatchId:         batch.Id,
+			TradableAmount:  batchSupplyTradable.String(),
+			RetiredAmount:   batchSupplyRetired.String(),
+			CancelledAmount: batchSupply.CancelledAmount,
+		}); err != nil {
+			return err
+		}
+		if err = sdk.UnwrapSDKContext(ctx).EventManager().EmitTypedEvent(&ecocredit.EventRetire{
+			Retirer:    to.String(),
+			BatchDenom: credit.BatchDenom,
+			Amount:     sendAmtRetired.String(),
+			Location:   credit.RetirementLocation,
+		}); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func getNonNegativeFixedDecs(precision uint32, decimals ...string) ([]math.Dec, error) {
+	decs := make([]math.Dec, len(decimals))
+	for i, decimal := range decimals {
+		dec, err := math.NewNonNegativeFixedDecFromString(decimal, precision)
+		if err != nil {
+			return nil, err
+		}
+		decs[i] = dec
+	}
+	return decs, nil
+}
+
+func (s serverImpl) getCreditTypeFromBatchDenom(ctx context.Context, denom string) (v1beta1.CreditType, error) {
+	classId := ecocredit.GetClassIdFromBatchDenom(denom)
+	classInfo, err := s.stateStore.ClassInfoStore().GetByName(ctx, classId)
+	if err != nil {
+		return v1beta1.CreditType{}, err
+	}
+	return s.getCreditType(sdk.UnwrapSDKContext(ctx), classInfo.CreditType)
+}
+
+func (s serverImpl) getClassSequenceNo(ctx context.Context, ctype string) (uint64, error) {
+	var seq uint64
+	classSeq, err := s.stateStore.ClassSequenceStore().Get(ctx, ctype)
+	switch err {
+	case nil:
+		seq = classSeq.NextClassId
+	case ormerrors.NotFound:
+		seq = 1
+	default:
+		return 0, err
+	}
+	err = s.stateStore.ClassSequenceStore().Save(ctx, &ecocreditv1beta1.ClassSequence{
+		CreditType:  ctype,
+		NextClassId: seq + 1,
+	})
+	return seq, err
+}
+
+func (s serverImpl) getCreditType(ctx sdk.Context, creditTypeName string) (v1beta1.CreditType, error) {
+	creditTypes := s.getAllCreditTypes(ctx)
+	creditTypeName = ecocredit.NormalizeCreditTypeName(creditTypeName)
+	for _, creditType := range creditTypes {
+		// credit type name's stored via params have enforcement on normalization, so we can be sure they will already
+		// be normalized here.
+		if creditType.Name == creditTypeName {
+			return *creditType, nil
+		}
+	}
+	return v1beta1.CreditType{}, sdkerrors.ErrInvalidType.Wrapf("%s is not a valid credit type", creditTypeName)
+}
+
+func (s serverImpl) getAllCreditTypes(ctx sdk.Context) []*v1beta1.CreditType {
+	var params ecocredit.Params
+	s.paramSpace.GetParamSet(ctx, &params)
```

Comment:
credit types we definitely don't want in params. it should just be in the `CreditType` table. i would say let's drop all param stuff now and add gov support later

Reply:
switched to use table